lib/hashie/mash.rb

```diff
@@ -121,20 +121,20 @@ class << self; alias [] new; end
     alias regular_reader []
     alias regular_writer []=
 
-    # Retrieves an attribute set in the Mash. Will convert
-    # any key passed in to a string before retrieving.
+    # Retrieves an attribute set in the Mash. Will convert a key passed in
+    # as a symbol to a string before retrieving.
     def custom_reader(key)
       default_proc.call(self, key) if default_proc && !key?(key)
       value = regular_reader(convert_key(key))
       yield value if block_given?
       value
     end
 
-    # Sets an attribute in the Mash. Key will be converted to
-    # a string before it is set, and Hashes will be converted
-    # into Mashes for nesting purposes.
+    # Sets an attribute in the Mash. Symbol keys will be converted to
+    # strings before being set, and Hashes will be converted into Mashes
+    # for nesting purposes.
     def custom_writer(key, value, convert = true) #:nodoc:
-      key_as_symbol = (key = convert_key(key)).to_sym
+      key_as_symbol = (key = convert_key(key)).respond_to?(:to_sym) ? key.to_sym : key
```

Comment:
Doesn't `convert_key` already do this? Or is it for when `convert_key` gets overridden?

Reply:
so, convert key is actually converting it to a string, not a symbol (i.e., when it comes in as a symbol). Convert key checks to whether it responds to to_sym so it knows whether its a case where it _should_ be converting to a string... Here we actually want it as a symbol if possible. It's confusing, I agree. Let me think for a sec and also see why we need it as a symbol at all. I don't like the `key = convert_key(key)` as part of the the expression in general either.

Reply:
Ah yes. Maybe write this on 2 lines to make it clearer?

Reply:
first one is weird cuz `converted_key` would be nil when it doesn't respond to `:to_sym` and we need the symbolized version for the `log_built_in_message` but the stringified version for `regular_writer` when it does respond to `to_sym` so we'd end up calling `convert_key(key)` twice.

```rb
def custom_writer(key, value, convert = true) #:nodoc:
  converted_key = convert_key(key)
  converted_key = converted_key.to_sym if converted_key.respond_to?(:to_sym)

  log_built_in_message(converted_key) if log_collision?(converted_key)
  regular_writer(convert_key(key), convert ? convert_value(value) : value)
end
```

second one i'm wary of because now we have two cases of methods that claim to be returning the input in some changed form while `convert_key` may not convert anything, and `key_as_symbol` would return either the input unchanged or `nil` in the case where it can't be symbolized, depending how it was written. Both options seem unexpected so maybe better to keep that bit of weirdness internal to this method.

Best I can think are these

```rb
def custom_writer(key, value, convert = true)
  key = convert_key(key)
  key_as_symbol = key.to_sym if key.respond_to?(:to_sym)

  log_built_in_message(key_as_symbol) if log_collision?(key_as_symbol)
  regular_writer(key, convert ? convert_value(value) : value)
end
```

```rb
def custom_writer(key, value, convert = true)
  log_built_in_message(key.to_sym) if key.respond_to?(:to_sym) && log_collision?(key.to_sym)
  regular_writer(convert_key(key), convert ? convert_value(value) : value)
end
```

opinion?

Reply:
i think the second is better/OK even with the repetitive `key.to_sym`... it's clear to me that the logging concern is only when the input could be used as a method call vs the guaranteed check in the first version but with a sometimes inappropriately named variable.